File: sdk/node/src/button.ts
import { v4 as uuidv4 } from 'uuid';
import { UIBuilder } from './uibuilder';
import { ButtonState, WidgetTypeButton } from './internal/session/state/button';
import { ButtonOptions } from './internal/options';

/**
 * Button component options
 */
export interface ButtonComponentOptions {
  /**
   * Whether the button is disabled
   * @default false
   */
  disabled?: boolean;
}

/**
 * Button component class
 */
export class Button {
  /**
   * Create a disabled button
   * @param disabled Whether the button is disabled
   * @returns Button options
   */
  static disabled(disabled: boolean): ButtonComponentOptions {
    return { disabled };
  }
}

/**
 * Add a button to the UI
 * @param builder The UI builder
 * @param label The button label
 * @param options Button options
 * @returns Whether the button was clicked
 */
export function button(
  builder: UIBuilder,
  label: string,
  options: ButtonComponentOptions = {},
): boolean {
  const runtime = builder.runtime;
  const session = builder.session;
  const page = builder.page;
  const cursor = builder.cursor;

  if (!session || !page || !cursor) {
    return false;
  }

  const buttonOpts: ButtonOptions = {
    label,
    disabled: options.disabled ?? false,
  };

  const path = cursor.getPath();
  const widgetID = builder.generatePageID(WidgetTypeButton, path);

  let buttonState = session.state.getButton(widgetID);
  if (!buttonState) {
    buttonState = new ButtonState(
      widgetID,
      false,
      buttonOpts.label,
      buttonOpts.disabled,
    );
  } else {
    buttonState.label = buttonOpts.label;
    buttonState.disabled = buttonOpts.disabled;
  }
  session.state.set(widgetID, buttonState);

  const buttonProto = convertStateToButtonProto(buttonState as ButtonState);
  runtime.wsClient.enqueue(uuidv4(), {
    sessionId: session.id,
    pageId: page.id,
    path: convertPathToInt32Array(path),
    widget: {
      id: widgetID,
      type: 'Button',
      button: buttonProto,
    },
  });

  cursor.next();

  return buttonState.value;
}

/**
 * Convert button state to proto
 * @param state Button state
 * @returns Button proto
 */
function convertStateToButtonProto(state: ButtonState): any {
  return {
    value: state.value,
    label: state.label,
    disabled: state.disabled,
  };
}

/**
 * Convert button proto to state
 * @param id Widget ID
 * @param data Button proto
 * @returns Button state
 */
export function convertButtonProtoToState(
  id: string,
  data: any,
): ButtonState | null {
  if (!data) {
    return null;
  }

  return new ButtonState(id, data.value, data.label, data.disabled);
}

/**
 * Convert path to int32 array
 * @param path Path
 * @returns Int32 array
 */
export function convertPathToInt32Array(path: number[]): number[] {
  return path.map((v) => v);
}

File: sdk/go/button.go
package sourcetool

import (
	"github.com/gofrs/uuid/v5"

	"github.com/trysourcetool/sourcetool-go/button"
	"github.com/trysourcetool/sourcetool-go/internal/options"
	websocketv1 "github.com/trysourcetool/sourcetool-go/internal/pb/websocket/v1"
	widgetv1 "github.com/trysourcetool/sourcetool-go/internal/pb/widget/v1"
	"github.com/trysourcetool/sourcetool-go/internal/session/state"
)

func (b *uiBuilder) Button(label string, opts ...button.Option) bool {
	buttonOpts := &options.ButtonOptions{
		Label:    label,
		Disabled: false,
	}

	for _, o := range opts {
		o.Apply(buttonOpts)
	}

	sess := b.session
	if sess == nil {
		return false
	}
	page := b.page
	if page == nil {
		return false
	}
	cursor := b.cursor
	if cursor == nil {
		return false
	}
	path := cursor.getPath()

	widgetID := b.generatePageID(state.WidgetTypeButton, path)
	buttonState := sess.State.GetButton(widgetID)
	if buttonState == nil {
		buttonState = &state.ButtonState{
			ID:    widgetID,
			Value: false,
		}
	}
	buttonState.Label = buttonOpts.Label
	buttonState.Disabled = buttonOpts.Disabled
	sess.State.Set(widgetID, buttonState)

	button := convertStateToButtonProto(buttonState)
	b.runtime.wsClient.Enqueue(uuid.Must(uuid.NewV4()).String(), &websocketv1.RenderWidget{
		SessionId: sess.ID.String(),
		PageId:    page.id.String(),
		Path:      convertPathToInt32Slice(path),
		Widget: &widgetv1.Widget{
			Id: widgetID.String(),
			Type: &widgetv1.Widget_Button{
				Button: button,
			},
		},
	})

	cursor.next()

	return buttonState.Value
}

func convertStateToButtonProto(state *state.ButtonState) *widgetv1.Button {
	return &widgetv1.Button{
		Value:    state.Value,
		Label:    state.Label,
		Disabled: state.Disabled,
	}
}

func convertButtonProtoToState(id uuid.UUID, data *widgetv1.Button) *state.ButtonState {
	if data == nil {
		return nil
	}
	return &state.ButtonState{
		ID:       id,
		Value:    data.Value,
		Label:    data.Label,
		Disabled: data.Disabled,
	}
}
